api/ombpdf/semdom.py

```diff
@@ -0,0 +1,186 @@
+from xml.dom import minidom
+
+from . import semtree
+
+
+class ListInfo:
+    def __init__(self, writer, info):
+        self.writer = writer
+        self.info = info
+        self.item = 1
+
+    def begin_list_item(self):
+        if self.info.is_ordered:
+            attrs = {
+                'emblem': str(self.item),
+                'marker': f'{self.item}.'
+            }
+        else:
+            attrs = {'marker': '\u2022'}
+        self.writer._push_child('listitem', attrs)
+        self.writer._push_child('para')
+        self.writer._push_child('content', elinfo=self)
+        self.item += 1
+
+    def end_list_item(self):
+        self.writer._pop_child('content')
+        self.writer._pop_child('para')
+        self.writer._pop_child('listitem')
+
+
+class DOMWriter(semtree.Writer):
+    def __init__(self):
+        self.muted = True
+        self.elinfo = {}
+        self.footnote_citations = {}
+        self.cursors = []
+
+    @property
+    def depth(self):
+        node = self.cursor
+        depth = 0
+        while node != self.root:
+            depth += 1
+            node = node.parentNode
+        return depth
+
+    def _push_cursor(self, cursor):
+        self.cursors.append(self.cursor)
+        self.cursor = cursor
+
+    def _pop_cursor(self):
+        self.cursor = self.cursors.pop()
+
+    def _push_child(self, name, attrs=None, elinfo=None):
+        child = self.document.createElement(name)
+        if attrs is not None:
+            for key, value in attrs.items():
+                child.setAttribute(key, value)
+        if elinfo is not None:
+            self.elinfo[child] = elinfo
+        self.cursor.appendChild(child)
+        self.cursor = child
+        return child
+
+    def _pop_child(self, name=None):
+        if name is not None:
+            assert self.cursor.nodeName == name
+        self.cursor = self.cursor.parentNode
+
+    def _get_parent(self, name):
+        node = self.cursor.parentNode
+        while node.nodeName != name:
+            node = node.parentNode
+        return node
+
+    def _ensure_section(self):
+        if self.cursor == self.root:
+            self._push_child('sec')
+
+    def _add_text(self, text):
+        self.cursor.appendChild(self.document.createTextNode(text))
+
+    def _add_comment(self, text):
+        self.cursor.appendChild(self.document.createComment(text))
+
+    # Below are Writer methods.
+
+    def begin_document(self, doc):
+        self.document = minidom.parseString('<policy></policy>')
+        self.root = self.document.childNodes[0]
+        self.cursor = self.root
+
+    def end_document(self, doc):
+        pass
+
+    def begin_heading(self, heading):
+        while self.depth >= heading.level:
+            self._pop_child()
+        self._push_child('sec', {'title': ''})
```

Comment:
I misunderstood when reading this and assumed the `title` attr wasn't ever set. Given that it isn't set until `create_text`, should we even mention it here?

Reply:
Oh, I had originally done that so I could use `+=` on the attribute's value without getting an "I can't append `None` to a string" error from Python, but I'll see if we can get away without that.